drivers/storage/portworx/status.go

```diff
@@ -415,9 +541,61 @@ func (p *portworx) updateRemainingStorageNodes(
 	return nil
 }
 
+// getNodesToPortworxPods find all k8s nodes where Portworx pods are running
+func (p *portworx) getNodesToPortworxPods(
+	cluster *corev1.StorageCluster,
+) (map[string][]*v1.Pod, error) {
+	pxLabels := p.GetSelectorLabels()
+	portworxPodList := &v1.PodList{}
+	err := p.k8sClient.List(
+		context.TODO(),
+		portworxPodList,
+		&client.ListOptions{
+			Namespace:     cluster.Namespace,
+			LabelSelector: labels.SelectorFromSet(pxLabels),
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get list of portworx pods. %v", err)
+	}
+
+	// Create a map of node name to px pod list
+	nodesToPods := make(map[string][]*v1.Pod)
+	for _, pod := range portworxPodList.Items {
+		controllerRef := metav1.GetControllerOf(&pod)
+		if controllerRef != nil && controllerRef.UID == cluster.UID && len(pod.Spec.NodeName) != 0 {
+			nodesToPods[pod.Spec.NodeName] = append(nodesToPods[pod.Spec.NodeName], pod.DeepCopy())
+		}
+	}
+	return nodesToPods, nil
+}
+
+// needsStorageNodeUpgrade checks if a storage node needs upgrade, there are 3 cases:
+// 1. check if storage node is updated according to px pod hash
+// 2. check if storage node is updated according to StorageCluster hash
+// 3. hash is updated but still needs node upgrade
+func (p *portworx) needsStorageNodeUpgrade(
+	cluster *corev1.StorageCluster,
+	storageNode *corev1.StorageNode,
+	portworxPod *v1.Pod,
+	hash string,
+) bool {
+	// storage cluster hash will override pod hash to check overall upgrade progress
+	expectedHash := portworxPod.Labels[util.DefaultStorageClusterUniqueLabelKey]
+	if hash != "" {
```

Comment:
This will mostly likely always be non-nil and will use the latest storagecluster's hash, right? Would that be a problem if only sub-set of nodes are supposed to update and now the pod hash and cluster hash don't match but technically all nodes are at their latest states?